### PR TITLE
CMS Types

### PR DESCRIPTION
**Pull request by joschka**
* getting data from the strapi file dump does now respect a given locale
* many UI components now have a zod schema for their props
* CMS types now have all zod schemas, also renamed things
* simplified the cms service and added an unified "api" (getStrapiPage, getStrapiFooter, ...) where the return values are typed
* Changed the "importing" of content.json, to a static import instead of using fs.readFile (simpler and better performance; only downside: content.json must exist, at least with an empty object inside
* the is now a translation service "props" to translate cms data to component props. Allows components to have another props structure than what comes from Strapi, good example: Image component

### CODE REVIEW

File: app/services/cms/models/BackgroundColor.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { z } from "zod";
+
+export const BackgroundColorSchema = z.union([
+  z.literal("default"),
+  z.literal("white"),
+  z.literal("blue"),
+  z.literal("yellow"),
+  z.literal("green"),
+  z.literal("red"),
+]);
```
Comment (chohner) [inline app/services/cms/models/BackgroundColor.ts:10]: nit: Maybe easier expressed with an enum  `z.enum(["default", "white", ....]);`?

File: app/services/cms/models/Background.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+import { WrapperSchema } from "./Wrapper";
+
+export const BackgroundSchema = WrapperSchema.merge(
+  z.object({
+    id: z.number(),
+    __component: z.literal("meta.background"),
+  })
+);
+
+export type Background = z.infer<typeof WrapperSchema>;
```
Comment (chohner) [inline app/services/cms/models/Background.ts:11]: this should export `BackgroundSchema`?

File: app/services/cms/models/Container.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+import { WrapperSchema } from "./Wrapper";
+
+export const ContainerSchema = WrapperSchema.merge(
+  z.object({
+    id: z.number(),
+    __component: z.literal("meta.container"),
+  })
+);
+
+export type Container = z.infer<typeof WrapperSchema>;
```
Comment (chohner) [inline app/services/cms/models/Container.ts:11]: `ContainerSchema`?

File: app/services/cms/models/ElementWithId.ts [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
-};
+export const ElementWithIdSchema = LocalizableSchema.merge(
+  TimestampableSchema
+).merge(
+  z.object({
+    id: z.number().optional(),
+    elementId: z.string(),
+    element: z.array(FormContentCmsSchema),
+  })
+);
```
Comment (chohner) [inline app/services/cms/models/ElementWithId.ts:14]: nit: I think merging the other way around might be easier to read?

```
export const ElementWithIdSchema =  z.object({
    id: z.number().optional(),
    elementId: z.string(),
    element: z.array(FormContentCmsSchema),
  }).merge(LocalizableSchema).merge(TimestampableSchema);
```

File: app/services/cms/models/Background.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { z } from "zod";
+import { WrapperSchema } from "./Wrapper";
+
+export const BackgroundSchema = WrapperSchema.merge(
+  z.object({
+    id: z.number(),
+    __component: z.literal("meta.background"),
+  })
+);
```
Comment (chohner) [inline app/services/cms/models/Background.ts:9]: nit: for me `z.object({}).merge(WrapperSchema)` might express this type more clear

File: app/services/cms/models/Heading.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  tagName: z.union([
+    z.literal("h1"),
+    z.literal("h2"),
+    z.literal("h3"),
+    z.literal("h4"),
+    z.literal("h5"),
+    z.literal("h6"),
+    z.literal("p"),
+    z.literal("div"),
+  ]),
```
Comment (chohner) [inline app/services/cms/models/Heading.ts:16]: nit: same here and below - `z.enum` might be a bit more ergonomic

File: app/services/cms/models/Localizable.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import { z } from "zod";
+import { LocaleSchema } from "./Locale";
+
+export const LocalizableSchema = z.object({
```
Comment (chohner) [inline app/services/cms/models/Localizable.ts:4]: nit: I like the zod naming of `HasX` (so here: `HasLocale`)

File: app/services/cms/models/BackgroundColor.ts (hunk 1 above)
Comment (joschka) [reply app/services/cms/models/BackgroundColor.ts:10]: Absolutely. I guess I mixed it up with TS enums in my mind.

File: app/services/cms/models/ElementWithId.ts (hunk 4 above)
Comment (joschka) [reply app/services/cms/models/ElementWithId.ts:14]: True

File: app/services/cms/models/Localizable.ts (hunk 7 above)
Comment (joschka) [reply app/services/cms/models/Localizable.ts:4]: Nice, way better than ---able.

File: app/components/Box.tsx [hunk 8]
```diff
@@ -1,16 +1,17 @@
...
+import { z } from "zod";
 
-export interface BoxProps {
-  label?: HeadingProps;
-  heading?: HeadingProps;
-  content?: ParagraphProps;
-  button?: ButtonProps;
-}
+export const BoxPropsSchema = z.object({
+  label: HeadingPropsSchema.optional(),
```
Comment (chohner) [inline app/components/Box.tsx:8]: `label` & `HeadingPropsSchema`?

File: app/services/cms/index.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+const getStrapiEntryFromSource =
+  config().CMS === "FILE" ? getStrapiEntryFromFile : getStrapiEntryFromApi;
+
+const getStrapiEntry = async (
+  opts: Pick<GetStrapiEntryOpts, "apiId" | "slug"> & {
+    locale?: StrapiLocale;
+  }
+) => {
+  const data = await getStrapiEntryFromSource({ locale: "de", ...opts });
+  // "remove attributes key"
```
Comment (frederike-ramin) [inline app/services/cms/index.ts:27]: Nit-picky: Why is the comment in quotes? Is there a specific reason/ practice that I'm just no aware of? :D

File: app/services/cms/models/StrapiBackground.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { z } from "zod";
+import { HasStrapiIdSchema } from "./HasStrapiId";
+import { StrapiWrapperSchema } from "./StrapiWrapper";
+
+export const StrapiBackgroundSchema = z
+  .object({
+    id: z.number(),
+    __component: z.literal("meta.background").optional(),
```
Comment (frederike-ramin) [inline app/services/cms/models/StrapiBackground.ts:8]: Why is the component literal optional? I think for a Background it should always be given? (same for the other components of course)

**Review by frederike-ramin — APPROVED** [review]
Nice. Looks really good, thank you for cleaning that up - feels a lot more stable that way. 👍

File: app/services/cms/index.ts (hunk 9 above)
Comment (joschka) [reply app/services/cms/index.ts:27]: I couldn't find a better wording. I want to say that I kind of remove the attributes key "in between".
`{ bla: { attributes: foobar } }`  => `{ bla: foobar }`. Not removing the whole key with value (`{ bla: {} }`). Remove a layer in between. Any idea?

Comment (frederike-ramin) [reply app/services/cms/index.ts:27]: Nono, I got that part. But shouldn't it be 
`  // remove "attributes" key`
then?

File: app/services/cms/models/StrapiBackground.ts (hunk 10 above)
Comment (joschka) [reply app/services/cms/models/StrapiBackground.ts:8]: :man_shrugging: sometimes it's there, sometimes not


kidding


Components that are nested inside another component don't have the `__component` field in the api response. For example: Heading at the top level inside content area of a page -> has `__component` field. Heading inside InfoBox -> no `__component` field.

File: app/components/form/steps/kinderAnzahl.tsx [hunk 11]
```diff
@@ -25,7 +25,9 @@ export const kinderAnzahlStep = {
             defaultValue={defaultValues?.[inputElement.name] || "0"}
             label={inputElement.label}
             type={inputElement.type}
-            placeholder={inputElement.placeholder}
+            placeholder={
+              inputElement.placeholder ? inputElement.placeholder : undefined
```
[the remark this answers is not in the document]
Comment (joschka) [reply app/components/form/steps/kinderAnzahl.tsx:29]: And `null`. In cms data, optional things are `null` when not set. But they are defined. In component props world optional things are expected to be `undefined`. Translation from `null` to `undefined`.

[the remark this answers is not in the document]
Comment (frederike-ramin) [reply app/components/form/steps/kinderAnzahl.tsx:29]: But I thought we're already getting rid of null during the strapi -> props translation?

File: app/services/cms/index.ts (hunk 9 above)
Comment (chohner) [reply app/services/cms/index.ts:27]: How about adding your example there? just so someone else doesn't need to dig through this PR :)

Comment (joschka) [reply app/services/cms/index.ts:27]:
```
// remove "attributes" key
// see https://github.com/digitalservicebund/a2j-rechtsantragstelle/pull/118 for more information
```

File: app/components/form/steps/kinderAnzahl.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (joschka) [reply app/components/form/steps/kinderAnzahl.tsx:29]: I changed my mind. Lets keep it for now and make it nicer when we refactor the steps.

**Pull request merged**